Refactor for plugin capability, add pods and npm clean commands

mikehardy opened the pull request:
Hi there!

A few of us in the react-native-community were discussing how we could never escape the need to completely wipe out state and start clean, and the react-native command line should just have a command for it.

The CLI maintainers mentioned we could probably do it as a plugin pretty quickly, and I found this project, and thought you might be okay with refactoring it so it could be used as a plugin?

So I did that, and I also added a couple extra possible commands - pod delete, pod update, and npm cache clean

It should work with both CLI 1.x and the upcoming 2.x for react-native 0.60, and you just need to do `react-native clean-project` or `react-native clean-project-auto` to have it either take you through the questions, or just purge everything respectively

If it's merge-worthy, it fixes https://github.com/react-native-community/cli/issues/387

pmadruga commented:
Thanks for this and yes, it can totally be used as a plugin. I will only ask for retroactive compatibility and any changes requested that may have come from the review. Jumping to the review now.

pmadruga commented on package.json:
```diff
@@ -8,6 +8,8 @@
   "main": "./source/index.js",
   "bin": "./source/index.js",
   "scripts": {
+    "create-example": "react-native init example && cd example && yarn add -D react-native-clean-project",
```
Just curious on why is this necessary?

mikehardy replied:
this was necessary because I saw a test entry in there but couldn't see how you were doing tests in here, so I went with ad-hoc testing using a quick example app I could sync changes into and run. Not automated but it is a very quick test/debug cycle, I must have run it 50 times while slowly making changes.

You mention tapejs, would you be willing to put an example of how you want to test things into master, then I can copy that pattern and exercise the changes?

Or if it is already there and I'm mixing something obvious just clue me in and I'll get to it

pmadruga replied:
You're right, there isn't a setup to run tests yet (sorry for that). Let me setup it and put it in master. I'll go with jest instead. 

I'll write here as soon as it's done.

pmadruga replied:
@mikehardy setup for running tests and a small test is now on master.

mikehardy replied:
Fantastic, I'll get after it today

pmadruga commented on source/plugin.js:
```diff
@@ -0,0 +1,27 @@
+module.exports = [
+  {
+    name: 'clean-project-auto',
```
The explanation for the why of these two options is necessary in the README.

mikehardy replied:
Totally agreed - I just wanted to see if you receptive before I did the administrivia

thymikee replied:
please add a `description` to the tasks for more helpful `--help` output of the CLI

mikehardy replied:
fantastic - I'll get to it, thanks!

pmadruga commented on source/internals.js:
```diff
@@ -0,0 +1,242 @@
+const { spawn } = require('child_process');
```
This file became rather big, I would rather create an 'internals' folder with the tasksList, the ask* and the gets into its own files.

mikehardy replied:
I would agree with that as well - it's pretty cluttered right now, but again I wanted to do something more minimal, like a Proposal Request more, first, since you're receptive I'll take your advice and re-shape this

mikehardy commented:
Oh great - I'm sure we can get this in a shape you like then. I'm pretty agnostic on everything *except* that things should work and be easy to maintain :-). I'll go through your review now

mikehardy commented:
Worth mentioning that I agree 100% on backwards compatibility, I see no reason this should be a breaking change in any way, it's really just making the code so it can be wrapped, then doing so. I think the community will really like this - they were excited when I showed them this package, since the UI and feedback is really cool for users

mikehardy commented:
Okay @pmadruga (and @thymikee if you have interest) - I think this one is an actual Pull Request as opposed to a quick-and-dirty Proposal Request like it was originally. Please take a fresh look and let me know what think?

mikehardy commented on CONTRIBUTING.md:
```diff
@@ -7,14 +7,14 @@ Please note we have a code of conduct, please follow it in all your interactions
+   variables, exposed ports, useful file locations and container parameters.
+1. Increase the version numbers in any examples files and the README.md to the new version that this
+   Pull Request would represent. The versioning scheme we use is [SemVer](http://semver.org/).
+1. You may merge the Pull Request in once you have the sign-off of two other developers, or if you
```
All the changes in CONTRIBUTING were just markdown lint

mikehardy replied:
Dang - must have to do with spacing / newlines. In markdown for numbered lists just repeat 1. and it works, assuming you do it right :-). I was just linting anyway - no good deed unpunished. I'll fix it

mikehardy commented on package.json:
```diff
@@ -8,6 +8,8 @@
   "bin": "./source/index.js",
   "scripts": {
+    "dev-init": "npx react-native init example",
+    "dev-sync": "cp -rv package.json source react-native.config.js example/node_modules/react-native-clean-project/",
```
Even with the automated testing, I took tremendous comfort in having quick access to a rebuildable example project with quick install of local module changes, so I left the example items in (here, .gitignore, and ignored in jest config). If you feel strongly that you don't like them, I can drop them

pmadruga commented on package.json:
```diff
@@ -27,7 +29,10 @@
     "jest": "^24.8.0"
+  },
+  "rnpm": {
+    "plugin": "./source/plugin.js"
```
I guess it's normal boilerplate to have the './source/plugin' reference both here and in the react-native.config.js. Should they always match?

mikehardy replied:
Yes - one is for react-native-community/cli v1 (the rnpm config) and one is for v2 (where they migrate to react-native.config.js. I wanted it to work now and in the pending release so I included the right hooks for both. I don't believe we can do comments in JSON or I'd leave a comment to remove the rnpm implementation later (once CLI v1 is long gone) but that would be the idea

pmadruga commented on source/plugin.js:
```diff
@@ -0,0 +1,39 @@
+      console.log('Use \`./node_modules/.bin/react-native-clean-project\' for total control');
+      console.log('');
+
+      rlInterface.close(); // if we don't do this it hangs waiting for input
```
The tests' results also raise a warning regarding an open task that's keeping preventing jest from exiting, and it's related to `rlInterface`. You can check it by running `./node_modules/.bin/jest --detectOpenHandles`. Can you give it a look, please?

mikehardy replied:
I gave that one a long look and I'm not sure how to fix it well. My general take was that it was like that before - by which I mean structurally you have the readline opened immediately whether it will be used or not, then you close it in an unrelated area. My conversion to a plugin style didn't relate to that though it did mean I had to be careful to close it myself in plugin.js. Unfortunately even though I'm closing it there it isn't closing when run by jest.

It was late enough at night and I'd worked enough on it that after contemplating the "no fix" path (no real consequence) I let it go. It's an existing issue with the code that needs cleaning but it'll require re-organizing how readline is used in the options.js which are orthogonal to my change.

Can we let it go for this PR?

pmadruga replied:
You also have a point here and I'll take it in another PR. Thank you for explaining it thoroughly.

mikehardy commented on CONTRIBUTING.md:
```diff
@@ -119,7 +119,7 @@ that are not aligned to this Code of Conduct, or to ban temporarily or
 threatening, offensive, or harmful.
 
-### Scope
+### Code of Conduct Scope
```
This was markdown lint - there were two headings with identical contents, so generated header anchors won't work. So I just changed one slightly. It's just lint so I don't really care, but it's a valid technical change - if it's a poor stylistic change, easy to revert. What do you think?

pmadruga replied:
Nop, looks good. Thank you!

mikehardy commented on source/index.js:
```diff
@@ -1,204 +1,45 @@
+const { executeTask } = require('./internals/executor');
+const { tasks } = require('./internals/tasks');
+
+options.askiOS()
```
Okay - let me think about that one (to do it well) and address it

mikehardy replied:
Hmm - on review I think this is original structure as well - they were mixed to start - and is orthogonal to my changes. I need to draw boundaries or I'll be on this forever. If it was simple I'd go after it immediately but really it should be two phases fully separate / one for all asking, and one for all execution. It's sort of like that now but if you want to disentangle it's really going to be up to your taste and there is a large chance I won't do it the way you'll want to live with. Additionally there already were OS-specific tasks, but they weren't handled that way to begin with (brew is macOS only, pods are new and macOS only so I just followed the pattern of brew, watchman isn't on linux but is on windows and mac, there's no "choco" but that is recommended for windows etc, also some people use yarn some use npm, the package install should check for yarn.lock or package-lock.json). So there's lots of scope here to do it really right, but I followed existing pattern. I'd like to leave this for now if possible?

pmadruga replied:
You're right and I totally agree to leave it for now. I'll get to it in the future then.

(I hope my comment before didn't sound as criticism)

mikehardy replied:
:-), no worries - I love a good engineering debate, it makes great software. Just running out of time and wanted to fix the scope out of self-interest haha

mikehardy commented:
Thanks so much for the quick review, great comments and I'll fix what's obviously warm while waiting for your reply on the rest.

mikehardy commented:
Okay - I've addressed all the comments, even if for the 2 more substantial ones I was essentially pleading to keep them separate. Let me know what you think - thanks!

pmadruga commented:
Thanks for the replies, @mikehardy . The PR looks great and exposed some of the issues that I need to address. Thumbs up from me. I'll merge as soon as possible - would just like @thymikee 's thumbs up. 

And overall, thanks for this PR.

thymikee commented on source/plugin.js:
```diff
@@ -0,0 +1,39 @@
+        taskPromises.push(executeTask(task));
+      });
+
+      Promise.all(taskPromises).then(() => {
```
```suggestion
      return Promise.all(taskPromises).then(() => {
```

By returning the result from the `func` function CLI is able to intercept the error messages and know when the command succeeded/failed

mikehardy commented:
Fantastic! I'm not part of the CLI team (I just run around making changes on things that drive me crazy personally) but I am going to use the heck out of this thing now. If the CLI folks want to keep this out of core I could see this plugin being the de facto standard at which point you and your module will be #InternetFamous, which is fun :-). Cheers

mikehardy commented:
@thymikee I'm so glad you looked at this, I would have never known that helped the CLI, not knowing the internals. I committed your suggestion exactly as is

mikehardy commented:
@pmadruga Any chance of a release with this soon :pray: :smiley: - would love to recommend this as a simple npm install vs using a git master dependency

pmadruga commented:
Done :)
